dotnet/src/Experimental/Orchestration.Flow/Execution/ReActEngine.cs

```diff
@@ -206,42 +207,37 @@ internal async Task<string> InvokeActionAsync(ReActStep actionStep, string chatI
 
             foreach (var variable in actionContextVariables)
             {
-                contextVariables.Set(variable.Key, variable.Value);
+                contextVariables[variable.Key] = variable.Value;
             }
 
             this._logger?.LogDebug("Invoked {FunctionName}. Result: {Result}", targetFunction.Name, result.GetValue<string>());
 
             return result.GetValue<string>() ?? string.Empty;
         }
-        catch (Exception e) when (!e.IsCriticalException())
+        catch (Exception e) when (!e.IsNonRetryable())
         {
             this._logger?.LogError(e, "Something went wrong in action step: {0}.{1}. Error: {2}", targetFunction.PluginName, targetFunction.Name, e.Message);
             return $"Something went wrong in action step: {targetFunction.PluginName}.{targetFunction.Name}. Error: {e.Message} {e.InnerException?.Message}";
         }
     }
 
-    private ContextVariables CreateActionContextVariables(Dictionary<string, string> actionVariables, Kernel kernel, ContextVariables contextVariables)
+    private KernelArguments CreateActionKernelArguments(Dictionary<string, string> actionVariables, KernelArguments context)
     {
-        var actionContext = contextVariables.Clone();
+        var actionContext = new KernelArguments(context);
 
         foreach (var kvp in actionVariables)
         {
-            actionContext.Set(kvp.Key, kvp.Value);
+            actionContext[kvp.Key] = kvp.Value;
         }
 
         return actionContext;
     }
 
-    private KernelFunction ImportSemanticFunction(Kernel kernel, string functionName, string promptTemplate, PromptTemplateConfig config)
+    private KernelFunction CreateSemanticFunction(Kernel kernel, string functionName, PromptTemplateConfig config)
     {
         var factory = new KernelPromptTemplateFactory(kernel.LoggerFactory);
```

Comment:
No need to create this, it will be used by default.

I think you can eliminate this method.

Reply:
thanks, removed this method.